awscli/customizations/s3/utils.py

```diff
@@ -278,22 +281,72 @@ def create_warning(path, error_message, skip_file=True):
 
 def find_chunksize(size, current_chunksize):
     """
-    The purpose of this function is determine a chunksize so that
-    the number of parts in a multipart upload is not greater than
-    the ``MAX_PARTS``.  If the ``chunksize`` is greater than
-    ``MAX_SINGLE_UPLOAD_SIZE`` it returns ``MAX_SINGLE_UPLOAD_SIZE``.
+    The purpose of this function is determine a chunksize so that the number of
+    parts in a multipart upload is not greater than the ``MAX_PARTS``, and that
+    the chunksize is not less than the minimum chunksize.
+
+    :param size: The size of the file to upload
+    :param current_chunksize: The currently configured chunksize
+    :return: If the given chunksize is valid, it is returned. Otherwise a valid
+        chunksize is calculated and returned.
+    """
+    size = adjust_chunksize_for_max_parts(size, current_chunksize)
+    return adjust_chunksize_to_upload_limits(size)
```

Comment:
It's been a while since I've looked at this so I want to make sure I understand this.

`adjust_chunksize_for_max_parts` only ever scales up the chunk size based on max number of allowed parts

`adjust_chunksize_to_upload_limits` can increase or decrease the chunk size based on max part_size.

So it's possible that `adjust_chunksize_to_upload_limits` can decrease the part size to something that would have been rejected/modified by `adjust_chunksize_for_max_parts` correct?

And the reason this is ok is because we're relying on the fact that something up the stack will reject the transfer anyways because at that point you've likely exceeded the max allowed file size in S3?